src/transform/src/analysis/equivalences.rs

```diff
@@ -73,115 +77,165 @@ impl Analysis for Equivalences {
-                for (pos, expr) in scalars.iter().enumerate() {
-                    equivalences
-                        .classes
-                        .push(vec![MirScalarExpr::Column(input_arity + pos), expr.clone()]);
+                let mut equivalences = results.get(index - 1).unwrap().clone();
+                if let Some(equivalences) = &mut equivalences {
+                    let input_arity = depends.results::<Arity>().unwrap()[index - 1];
+                    for (pos, expr) in scalars.iter().enumerate() {
+                        equivalences
+                            .classes
+                            .push(vec![MirScalarExpr::Column(input_arity + pos), expr.clone()]);
+                    }
                 }
                 equivalences
             }
-            MirRelationExpr::FlatMap { .. } => results.last().unwrap().clone(),
+            MirRelationExpr::FlatMap { .. } => results.get(index - 1).unwrap().clone(),
             MirRelationExpr::Filter { predicates, .. } => {
-                let mut equivalences = results.last().unwrap().clone();
-                let mut class = predicates.clone();
-                class.push(MirScalarExpr::literal_ok(
-                    Datum::True,
-                    mz_repr::ScalarType::Bool,
-                ));
-                equivalences.classes.push(class);
+                let mut equivalences = results.get(index - 1).unwrap().clone();
+                if let Some(equivalences) = &mut equivalences {
+                    let mut class = predicates.clone();
+                    class.push(MirScalarExpr::literal_ok(
+                        Datum::True,
+                        mz_repr::ScalarType::Bool,
+                    ));
+                    equivalences.classes.push(class);
+                }
                 equivalences
             }
             MirRelationExpr::Join { equivalences, .. } => {
                 // Collect equivalences from all inputs;
-                let expr_index = results.len();
+                let expr_index = index;
                 let mut children = depends
                     .children_of_rev(expr_index, expr.children().count())
                     .collect::<Vec<_>>();
                 children.reverse();
 
                 let arity = depends.results::<Arity>().unwrap();
                 let mut columns = 0;
-                let mut result = EquivalenceClasses::default();
+                let mut result = Some(EquivalenceClasses::default());
                 for child in children.into_iter() {
                     let input_arity = arity[child];
-                    let mut equivalences = results[child].clone();
-                    let permutation = (columns..(columns + input_arity)).collect::<Vec<_>>();
-                    equivalences.permute(&permutation);
-                    result.classes.extend(equivalences.classes);
+                    let equivalences = results[child].clone();
```

Comment:
Can we avoid this clone and instead clone `equivalences.classes` below?

Reply:
Also, `equivalences` shadows a binding from the outer scope, maybe it's a good opportunity to rename?

Reply:
Ah, if I told you that `equivalences` has exactly one field, name `classes`, would that change your mind?

Is the shadowing you are worried about the `if let`? I'm used to re-using the name in that case, but if it's clearer changing one to `x` or something, can do.

Reply:
> Ah, if I told you that equivalences has exactly one field, named classes, would that change your mind?

Should we make it a tuple struct then? :) No strong feelings here!